// js_lis/layouts.js
export const layouts = {
  en: [
      ['`', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '=', 'Backspace'],
      ['Tab', 'q', 'w', 'e', 'r', 't', 'y', 'u', 'i', 'o', 'p', '[', ']', '\\'],
      ['Caps', 'a', 's', 'd', 'f', 'g', 'h', 'j', 'k', 'l', ';', '\'', 'Enter'],
      ['Shift', 'z', 'x', 'c', 'v', 'b', 'n', 'm', ',', '.', '/', 'Shift'],
      ['Space']
  ],
  th: [
      ['_', 'ๅ', '/', '-', 'ภ', 'ถ', 'ุ', 'ึ', 'ค', 'ต', 'จ', 'ข', 'ช', 'Backspace'],
      ['Tab', 'ๆ', 'ไ', 'ำ', 'พ', 'ะ', 'ั', 'ี', 'ร', 'น', 'ย', 'บ', 'ล', 'ฃ'],
      ['Caps', 'ฟ', 'ห', 'ก', 'ด', 'เ', '้', '่', 'า', 'ส', 'ว', 'ง', 'Enter'],
      ['Shift', 'ผ', 'ป', 'แ', 'อ', 'ิ', 'ื', 'ท', 'ม', 'ใ', 'ฝ', 'Shift'],
      ['Space']
  ],
  numpad: [
    ["+", "-", "*", "/"],
    ["1", "2", "3", "%"],
    ["4", "5", "6", "."],
    ["7", "8", "9", "="],
    ["00", "0", "backspace"],
  ],
  scNum : [
    ["+", "-", "*", "/"],
    ["1", "2", "3", "%"],
    ["4", "5", "6", "."],
    ["7", "8", "9", "="],
    ["00", "0", "backspace"],
  ],
  thSc : [
    ["ก", "ข", "ฃ", "ค", "ฅ", "ฆ", "ง", "จ", "ฉ", "ช", "ซ", "ฌ", "Backspace"],
    ["ญ", "ฎ", "ฏ", "ฐ", "ฑ", "ฒ", "ณ", "ด", "ต", "ถ", "ท", "ธ", "น"],
    ["บ", "ป", "ผ", "ฝ", "พ", "ฟ", "ภ", "ม", "ย", "ร", "ฤ", "Enter"],
    ["ล", "ฦ", "ว", "ศ", "ษ", "ส", "ห", "ฬ", "อ", "ฮ"],
    ["Space"],
  ],
  enSc: [
    ["`", "1", "2", "3", "4", "5", "6", "7", "8", "9", "0", "="],
    ["Tab", "q", "w", "e", "r", "t", "y", "u", "i", "o", "p", "Backspace"],
    ["Caps", "a", "s", "d", "f", "g", "h", "j", "k", "l", "Enter"],
    ["Shift", "z", "x", "c", "v", "b", "n", "m", "Shift"],
    ["Space"],
  ],
};
